src/uct/tcp/tcp_ep.c

```diff
@@ -69,7 +69,8 @@ static inline int uct_tcp_ep_ctx_buf_need_progress(uct_tcp_ep_ctx_t *ctx)
 static inline ucs_status_t uct_tcp_ep_check_tx_res(uct_tcp_ep_t *ep)
 {
     if (ucs_likely((ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED) &&
-                   uct_tcp_ep_ctx_buf_empty(&ep->tx))) {
+                   uct_tcp_ep_ctx_buf_empty(&ep->tx) &
+                   (ep->put_cnt != UINT32_MAX))) {
```

Comment:
looks like it will fail on assert below if `ep->put_cnt == UINT32_MAX` now, because of `ep->conn_state`

also do we really need this check for all ops?

Reply:
yes, we have to check it for all operations to avoid sending if something in the pending queue

Reply:
good catch, fixed

Reply:
can we assume puts will be unordered with other ops?

Reply:
@brminich is it ok?